spark-plugins/src/main/java/co/cask/hydrator/plugin/spark/SkipGramTrainer.java

```diff
@@ -0,0 +1,172 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.hydrator.plugin.spark;
+
+import co.cask.cdap.api.annotation.Description;
+import co.cask.cdap.api.annotation.Name;
+import co.cask.cdap.api.annotation.Plugin;
+import co.cask.cdap.api.data.format.StructuredRecord;
+import co.cask.cdap.api.data.schema.Schema;
+import co.cask.cdap.api.dataset.lib.FileSet;
+import co.cask.cdap.api.plugin.PluginConfig;
+import co.cask.cdap.etl.api.PipelineConfigurer;
+import co.cask.cdap.etl.api.batch.SparkExecutionPluginContext;
+import co.cask.cdap.etl.api.batch.SparkPluginContext;
+import co.cask.cdap.etl.api.batch.SparkSink;
+import com.google.common.base.Preconditions;
+import org.apache.spark.api.java.JavaRDD;
+import org.apache.spark.api.java.JavaSparkContext;
+import org.apache.spark.ml.feature.Word2Vec;
+import org.apache.spark.ml.feature.Word2VecModel;
+import org.apache.spark.sql.DataFrame;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.SQLContext;
+import org.apache.spark.sql.types.StructField;
+import org.apache.spark.sql.types.StructType;
+
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Set;
+import javax.annotation.Nullable;
+
+/**
+ * Spark Sink plugin that trains a model using SkipGram (Spark's Word2Vec).
+ * Saves this model to a FileSet.
+ */
+@Plugin(type = SparkSink.PLUGIN_TYPE)
+@Name(SkipGramTrainer.PLUGIN_NAME)
+@Description("Fits a model using SkipGram.")
+public class SkipGramTrainer extends SparkSink<StructuredRecord> {
+  public static final String PLUGIN_NAME = "SkipGramTrainer";
+  private SkipGramTrainerConfig config;
+  private StructType schema;
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    pipelineConfigurer.createDataset(config.fileSetName, FileSet.class);
+    Schema inputSchema = pipelineConfigurer.getStageConfigurer().getInputSchema();
+    Preconditions.checkArgument(inputSchema != null, "Input Schema must be a known constant.");
+    config.validate(inputSchema);
+  }
+
+  @Override
+  public void run(SparkExecutionPluginContext context, JavaRDD<StructuredRecord> input)
+    throws Exception {
+    if (input == null) {
+      return;
+    }
+
+    JavaSparkContext javaSparkContext = context.getSparkContext();
+    SQLContext sqlContext = new SQLContext(javaSparkContext);
+    Set<String> inputCols = new LinkedHashSet<>();
+    inputCols.add(config.inputCol);
+    List<StructField> structField = SparkUtils.getStructFieldList(input.first().getSchema().getFields(), inputCols);
+    schema = new StructType(structField.toArray(new StructField[structField.size()]));
+    JavaRDD<Row> rowRDD = SparkUtils.convertJavaRddStructuredRecordToJavaRddRows(input, schema);
```

Comment:
you don't need to use the DataFrame version and do all this conversion. Can just do something like:

```
    final String fieldName = config.inputCol;
    final Splitter splitter = config.getSplitt();
    JavaRDD<Iterable<String>> textRDD = input.map(new Function<StructuredRecord, Iterable<String>>() {
      @Override
      public Iterable<String> call(StructuredRecord input) throws Exception {
        String text = input.get(fieldName);
        return splitter.split(text);
      }
    });
    Word2Vec word2Vec = new Word2Vec()
     .setVectorSize(config.vectorSize)
     ...
    Word2VecModel model = word2Vec.fit(textRDD);
    model.save(...)

```